[WIP] ES6 module system

vincedani opened the pull request:
Initial commit of the ES6 module system.

JerryScript-DCO-1.0-Signed-off-by: Daniel Vince vinced@inf.u-szeged.hu

rerobika commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1644,6 +1661,172 @@ parser_parse_continue_statement (parser_context_t *context_p) /**< context */
+    case LEXER_KEYW_CLASS:
+    {
+      // TODO: This part is going to be implemented in the next part of the patch.
+      parser_parse_expression (context_p, PARSE_EXPR);
```
Throw unimplemented feature error!

rerobika commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1644,6 +1661,172 @@ parser_parse_continue_statement (parser_context_t *context_p) /**< context */
+    lexer_expect_object_literal_id (context_p, false);
+    // TODO: Import the requested properties from the given script and export
+    //       them from the current to make a redirection.
+    // This part is going to be implemented in the next part of the patch.
```
Ditto.

rerobika commented on jerry-core/parser/js/js-parser.c:
```diff
@@ -2291,19 +2291,20 @@ parser_parse_source (const uint8_t *arg_list_p, /**< function argument list */
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+                     ,
+                     parser_module_node_t *module_node_p, /**< module node that contains imports */
+                     parser_module_node_t *export_node_p  /**< module node that contains exports */
```
Uhh, this is a bit ugly! Cannot these values can be part of JERRY_CONTEXT, which could contain the current {import, export}_module_t? See how the JERRY_CONTEXT (vm_top_context) works.

rerobika commented on jerry-core/parser/js/js-parser.c:
```diff
@@ -2291,19 +2291,20 @@ parser_parse_source (const uint8_t *arg_list_p, /**< function argument list */
   context.error = PARSER_ERR_NO_ERROR;
   context.allocated_buffer_p = NULL;
 
-  if (error_location_p != NULL)
```
Why this code part is moved?

rerobika commented on jerry-core/parser/js/js-parser.c:
```diff
@@ -2928,6 +2958,126 @@ parser_parse_script (const uint8_t *arg_list_p, /**< function argument list */
 
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+/**
+ * Parse EcamScript module source code
```
Typo.

rerobika commented on jerry-core/parser/js/js-parser.c:
```diff
@@ -2928,6 +2958,126 @@ parser_parse_script (const uint8_t *arg_list_p, /**< function argument list */
+ *         syntax error - otherwise
+ */
+ecma_value_t
+parser_parse_module_script (const uint8_t *arg_list_p, /**< function argument list */
```
This function is almost the duplicate of `parser_parse_script`. Cannot these function can be merged somehow?

rerobika commented on jerry-core/vm/vm.c:
```diff
@@ -211,6 +212,32 @@ static const uint16_t vm_decode_table[] JERRY_CONST_DATA =
+ */
+ecma_value_t
+vm_run_module (const ecma_compiled_code_t *bytecode_p, /**< pointer to bytecode to run */
+              ecma_object_t *lex_env_p) /**< pointer to the specified lexenv to run in */
```
misaligned line.

galpeter commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1644,6 +1661,172 @@ parser_parse_continue_statement (parser_context_t *context_p) /**< context */
+  JERRY_ASSERT (context_p->token.type == LEXER_KEYW_IMPORT);
+  parser_module_context_init (context_p);
+
+  parser_module_node_t *import_node_p = parser_module_create_import_node (context_p);
```
Maybe this could be a local var instead.

galpeter commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1644,6 +1661,172 @@ parser_parse_continue_statement (parser_context_t *context_p) /**< context */
+      parser_parse_function_statement (context_p, function_name_p);
+
+      lexer_literal_t *import_name_p = function_name_p;
+      lexer_literal_t *local_name_p = function_name_p;
```
There is no need for the two local vars, just simly pass for the next method the `function_name_p`.

galpeter commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1644,6 +1661,172 @@ parser_parse_continue_statement (parser_context_t *context_p) /**< context */
+      parser_parse_var_statement (context_p, variable_name_p);
+
+      lexer_literal_t *import_name_p = variable_name_p;
+      lexer_literal_t *local_name_p = variable_name_p;
```
same as above

galpeter commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+} /* parser_module_partial_cleanup_on_error */
+
+static void
+parser_module_cleanup_export_node (parser_module_node_t *export_node_p)
```
IMHO: This and the method above could be mereged somehow.

galpeter commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+{
+  if (module_node_p != NULL)
+  {
+    parser_module_delete_saved_names (module_node_p->module_names,
```
We could simply pass the `module_node_p` directly and accessing the `names`, `module_request_count` can be done in that method.

galpeter commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+      parser_module_names_t *new_module_names = module_node_p->module_names;
+      is_stored_module = true;
+
+      for (uint8_t i = 0; i < module_node_p->module_request_count - 1 ; i++)
```
there is an extra whitespace before the last semicolon

galpeter commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+        new_module_names = new_module_names->next;
+      }
+
+      new_module_names->next=stored_imports->module_names;
```
please add whitespaces around the assignment.

galpeter commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+    ecma_string_t *import_name_p = ecma_new_ecma_string_from_utf8 (current->local_name_p,
+                                                                   current->local_name_length);
+
+    if (ecma_compare_ecma_strings (import_name_p, property_name_p))
```
this could be: ` bool found = ecma_compare_...()`

galpeter commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+
+    if (buffer_p  == NULL)
+    {
+      jerry_port_free_buffer (buffer_p);
```
In this case there is no need for the "free" call.

vincedani commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+  if (context_p->module_context_p->exports == NULL)
+  {
+    context_p->module_context_p->exports = module_node_p;
+  }
```
There is exactly one export node (`exports_p`) to each script which exports variables. This function creates the node when first reaches an `export statement`. The subsequent exports will be appended to this node.

vincedani commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+  if (jerry_value_is_error (ret_value))
+  {
+    jerry_release_value (ret_value);
+    parser_raise_error (context_p, PARSER_ERR_MODULE_REQUEST_NOT_FOUND);
```
Great idea, I've added this to the unsupported feature list and also to my TODO list for the second patch.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+  if (context_p->token.type == LEXER_LITERAL
+      && lexer_compare_raw_identifier_to_current (context_p, "from", 4))
+  {
+    parser_raise_error (context_p, PARSER_ERR_INVALID_CHARACTER);
```
If an error is raised, whether the initialized context is freed? (This should be done in the `PARSER_CATCH` part of js-parser.c

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+
+  if (context_p->token.type == LEXER_KEYW_DEFAULT)
+  {
+    // TODO: This part is going to be implemented in the next part of the patch.
```
Please do not use single line comments.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,735 @@
+    }
+    lexer_next_token (context_p);
+  }
+} /* parser_module_parse_import_item_list */
```
I am still voting for merging these functions.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+                                    parser_module_names_t *export_name_p) /**< export name */
+{
+  parser_module_names_t *current = export_node_p->module_names_p;
+  parser_module_names_t *next;
```
Use _p suffixes.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+  for (int i = 0; i < export_node_p->module_request_count; i++)
+  {
+    next = current->next_p;
+    ecma_string_t *import_name_p = ecma_new_ecma_string_from_utf8 (current->local_name_p,
```
All these string can be determined during parse time, therefor ecma_find_or_create_literal can be used here. (Maybe some modifications are required)

vincedani replied:
The first row of that [function](https://github.com/pando-project/jerryscript/blob/master/jerry-core/ecma/base/ecma-literal-storage.c#L73) is `ecma_string_t *string_p = ecma_new_ecma_string_from_utf8 (chars_p, size);` and does a lot more (unnecessary in this case) things after it, so I think this method is not as costly as calling the `ecma_find_or_create_literal_string`.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+                                                 parser_module_names_t *export_names_p, /**< export names */
+                                                 parser_module_names_t *eventual_names_p) /**< [out] used names */
+{
+  JERRY_UNUSED (eventual_names_p);
```
This is unnecessary.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+  }
+
+  return false;
+} /* parser_module_compare_property_name_with_import */
```
Isn't this function quite similar to the function above?

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+
+  if (ecma_get_object_type (func_obj_p) != ECMA_OBJECT_TYPE_FUNCTION)
+  {
+    parser_raise_error (context_p, PARSER_ERR_ARGUMENTS_NOT_ALLOWED);
```
Is this error message is correct?

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+  jerry_value_t func_val = ret_value;
+  ecma_object_t *func_obj_p = ecma_get_object_from_value (func_val);
+
+  if (ecma_get_object_type (func_obj_p) != ECMA_OBJECT_TYPE_FUNCTION)
```
What other object type could happen here?

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+  ecma_object_t *scope_p = ecma_create_decl_lex_env (ecma_get_global_environment ());
+  ret_value = vm_run_module (ecma_op_function_get_compiled_code ((ecma_extended_object_t *) func_obj_p), scope_p);
+
+  ecma_object_t *global_lex_env_p = ecma_get_lex_env_binding_object (ecma_get_global_environment ());
```
A lex_env's binding object cannot be a lex_env. The also the global lexenv's binding object is the builtin global object.

vincedani replied:
Rename to `global_obj_p` and use `ecma_object_t *glob_obj_p = ecma_builtin_get (ECMA_BUILTIN_ID_GLOBAL);`

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+                                                                             new_name.import_name_length);
+
+        ecma_create_named_data_property (global_lex_env_p, new_property_name_p,
+                                         ECMA_PROPERTY_NOT_WRITABLE,
```
New line after the second argument.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+    size_t size = 0;
+    uint8_t *buffer_p = jerry_port_module_read_source ((const char *) script_path, &size);
+
+    if (buffer_p  == NULL)
```
Only one space

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+
+    if (buffer_p  == NULL)
+    {
+      jerry_port_module_release_source (buffer_p);
```
If it is NULL, why should it be released?

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+    if (whole_module_needed)
+    {
+      local_name_p = NULL;
+      goto check_next_token;
```
Is that goto is needed? Cannot be substituted somehow?

rerobika commented on tools/run-tests.py:
```diff
@@ -353,6 +353,7 @@ def run_jerry_tests(options):
 
         if '--profile=es2015-subset' in job.build_args:
             skip_list.append(r"es5.1\/")
+            skip_list.append(r"es2015\/module-imported.js")
```
Why is this file is skipped? If the script exports variables/functions, but nobody imports them, is should be also a valid, leak-free test as well.

vincedani commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,803 @@
+
+      stored_imports->module_names_p = module_names_p;
+      stored_imports->module_request_count =
+      (uint8_t) (stored_imports->module_request_count + module_node_p->module_request_count);
```
Overflow check.

vincedani commented on tools/build.py:
```diff
@@ -140,6 +140,8 @@ def devhelp(helpstring):
                          help=devhelp('enable Valgrind support (%(choices)s)'))
     coregrp.add_argument('--vm-exec-stop', metavar='X', choices=['ON', 'OFF'], type=str.upper,
                          help='enable VM execution stopping (%(choices)s)')
+    coregrp.add_argument('--enable-module-system', metavar='X', choices=['ON', 'OFF'], type=str.upper,
```
The feature is disabled by default: [link](https://github.com/pando-project/jerryscript/pull/2599/files#diff-6f17516729bf6971897fbd29ea7e773cR42)

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,604 @@
+#include "module.h"
+
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+#define MAX_IMPORT_COUNT 65535
```
Use UINT16_MAX instead, also please add a documentation for it.

rerobika commented on .travis.yml:
```diff
@@ -6,7 +6,7 @@ dist: trusty
 # Default job task: run tests as defined in the $OPT environment variable.
 # Jobs can redefine the 'script' stage in the matrix below.
-script: tools/run-tests.py $OPTS
+script: travis_wait 20 tools/run-tests.py $OPTS
```
Is this really necessary?

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,604 @@
+        && context_p->token.type != LEXER_KEYW_DEFAULT
+        && (context_p->token.type != LEXER_LITERAL
+          || lexer_compare_raw_identifier_to_current (context_p, "from", 4)
+          || lexer_compare_raw_identifier_to_current (context_p, "as", 2)))
```
Wrong indentation.

rerobika commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -71,6 +71,11 @@ typedef enum
   PARSER_STATEMENT_TRY,
+
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+  PARSER_STATEMENT_IMPORT,
```
I think it's unnecessary.

rerobika commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -306,6 +311,7 @@ parser_parse_enclosed_expr (parser_context_t *context_p) /**< context */
 
 /**
  * Parse var statement.
+ * @return the name of the variable
```
This comment seems outdated to me.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,604 @@
+
+  while (stored_imports != NULL)
+  {
+    if (stored_imports->script_path_p == module_node_p->script_path_p)
```
You should use memcmp here.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,604 @@
+
+      if (memcmp (import_name_p->local_name_p,
+                     export_iterator_p->import_name_p,
+                     import_name_p->local_name_length) == 0)
```
Wrong indentation.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,604 @@
+} /* parser_module_check_request_place */
+
+void
+parser_module_handle_from_clause (parser_context_t *context_p)
```
Missing documentation for the function and the argument.

rerobika commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -362,6 +379,7 @@ parser_parse_var_statement (parser_context_t *context_p) /**< context */
 
 /**
  * Parse function statement.
+ * @return the name of the function
```
Ditto.

rerobika commented on docs/14.MODULE-SYSTEM.md:
```diff
@@ -0,0 +1,132 @@
+
+## General
+
+If the main script contains import statements, then Jerry open and run the appropriate scripts before the main script (as the standard says). The script's and the module's extension is `.js`, custom extensions are unnecessary.
```
open-> opens, run -> runs

rerobika commented on jerry-core/config.h:
```diff
@@ -50,6 +50,12 @@
+#if (defined (CONFIG_DISABLE_ES2015) || !defined (JERRY_ENABLE_MODULE_SYSTEM))
+# define CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+#elif (defined (CONFIG_DISABLE_ES2015) && defined (JERRY_ENABLE_MODULE_SYSTEM))
+#  error "ES2015 feature set is not configured correctly. The module system is ES2015 feature"
```
Please change it to: `Module system requires es2015 subset.`

rerobika commented on jerry-core/ecma/operations/ecma-lex-env.h:
```diff
@@ -34,6 +34,11 @@ void ecma_init_global_lex_env (void);
 ecma_object_t *ecma_get_global_environment (void);
 
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+void ecma_module_add_lex_env (ecma_object_t *);
```
Please add the corresponding argument name as well.

rerobika commented on jerry-core/jcontext/jcontext.h:
```diff
@@ -102,6 +114,12 @@ struct jerry_context_t
+
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
+  ecma_module_lex_envs_t *ecma_module_lex_envs_p; /**< list of module's lexical environments */
+  parser_module_context_t *module_top_context_p; /**< module context */
```
Change the comment to: `top (current) module parser context`

rerobika commented on jerry-core/module/module.c:
```diff
@@ -0,0 +1,265 @@
+} /* parser_module_compare_property_name_with_import */
+
+/**
+ * Connect the imported script's propeties to the main script.
```
Typo: `propeties` -> `properties`

rerobika commented on jerry-core/module/module.c:
```diff
@@ -0,0 +1,265 @@
+ * Connect the imported script's propeties to the main script.
+ */
+static void
+module_connect_properties (ecma_object_t *scope_p) /** scope */
```
`/** scope */` -> `/**< scope */`

rerobika replied:
Also add an assert for that `scope_p` should be declarative lexical environment.

rerobika commented on jerry-core/parser/js/js-parser-internal.h:
```diff
@@ -296,6 +296,61 @@ typedef struct
+ */
+typedef struct parser_module_names
+{
+  parser_module_utf8_string_t import_name; /**< local name of the import item */
```
isnt this structure used for export items as well? Please update the doc.

rerobika commented on jerry-core/parser/js/js-parser-internal.h:
```diff
@@ -296,6 +296,61 @@ typedef struct
+
+/**
+ * Module context in the parser context. It is not in the context if modules
+ * are not enabled during the build. It's value is NULL if there is no import /
```
I think the second sentence is unnecessary.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+    permanent_node_p->next_p = context_p->module_context_p->imports_p;
+    context_p->module_context_p->imports_p = permanent_node_p;
+  }
+
```
Unnecessary new line.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+
+/**
+ * Create import node.
+ * @return - the copy of the temlpate if the second parameter is not NULL.
```
typo: `temlpate`

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+      import_name_p = context_p->lit_object.literal_p;
+    }
+
+
```
Only one new line is enough

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+
+  module_node_p->script_path.length = (prop_length_t)(context_p->lit_object.literal_p->prop.length + 1);
+  module_node_p->script_path.value_p = (uint8_t *) parser_malloc (context_p,
+                                                            module_node_p->script_path.length * sizeof (uint8_t));
```
Wrong indentation.

rerobika commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+    parser_raise_error (context_p, PARSER_ERR_PROPERTY_IDENTIFIER_EXPECTED);
+  }
+
+  module_node_p->script_path.length = (prop_length_t)(context_p->lit_object.literal_p->prop.length + 1);
```
Add one space after the cast

rerobika commented on jerry-core/parser/js/js-parser-util.c:
```diff
@@ -1064,6 +1064,46 @@ parser_error_to_string (parser_error_t error) /**< error code */
+    }
+    case PARSER_ERR_STRING_EXPECTED:
+    {
+      return "Expected string token.";
```
`Expected string token.` -> `Expected a string.`

rerobika commented on jerry-core/parser/js/js-parser-util.c:
```diff
@@ -1064,6 +1064,46 @@ parser_error_to_string (parser_error_t error) /**< error code */
+    }
+    case PARSER_ERR_MODULE_UNEXPECTED:
+    {
+      return "Import and export statement must be in the global context.";
```
`statement` -> `statements`

rerobika commented on jerry-core/parser/js/js-parser.h:
```diff
@@ -129,7 +131,18 @@ typedef enum
-  PARSER_ERR_NON_STRICT_ARG_DEFINITION                /**< non-strict argument definition */
+  PARSER_ERR_NON_STRICT_ARG_DEFINITION,               /**< non-strict argument definition */
+
+#ifndef CONFIG_DISABLE_ES2015_MODULE_SYSTEM
```
Move these group before `PARSER_ERR_OBJECT_PROPERTY_REDEFINED`. Therefore the last enum member is not modified.

zherczeg commented on jerry-core/ecma/operations/ecma-lex-env.c:
```diff
@@ -68,6 +68,46 @@ ecma_get_global_environment (void)
+  JERRY_ASSERT (lex_env_p != NULL);
+  JERRY_ASSERT (ecma_is_lexical_environment (lex_env_p));
+
+  ecma_module_lex_envs_t *new_wrapper_p;
```
new_module_lex_env.

zherczeg commented on jerry-core/parser/js/js-parser-internal.h:
```diff
@@ -296,6 +296,59 @@ typedef struct
+typedef struct parser_module_names
+{
+  parser_module_utf8_string_t import_name; /**< local name of the import - export item */
+  parser_module_utf8_string_t local_name;  /**< import name of the import - export item */
```
Is this a local->import map? If yes, what about using objects for this (for faster property search).

vincedani replied:
Thanks for your review, I’ve updated the patch.

`import { a as b } from 'module.js';`

If a specific module (`module.js`) has been run, its lexical environment is being checked. If that lexical environment contains a property with the given name (`a`), then a new property is created in the parent context with the imported name (`b`) and assigned with the value of `a`.

If objects are being used for the local name -> import name mapping, the key should be the module path. This object should contain as many key-value pairs as many import items were in the import statement. 
![module_map](https://user-images.githubusercontent.com/16573361/53647705-63b5ab80-3c3e-11e9-8a44-d534bf28d01c.png)

I think in this scenario objects won’t worth it, because we now have 1 property search per import item (_in the lexical environment_), then we should have min. 2 (_in the object too_).

zherczeg commented on jerry-core/parser/js/js-parser.h:
```diff
@@ -90,6 +90,8 @@ typedef enum
   PARSER_ERR_LEFT_BRACE_EXPECTED,                     /**< left brace expected */
   PARSER_ERR_RIGHT_PAREN_EXPECTED,                    /**< right paren expected */
   PARSER_ERR_RIGHT_SQUARE_EXPECTED,                   /**< right square expected */
+  PARSER_ERR_LEFT_PAREN_MULTIPLY_LITERAL_EXPECTED,    /**< left paren or multiply or literal expected */
```
Isn't this module specific?

LaszloLango commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+
+    if (current_p != NULL && current_length == import_name_p->prop.length)
+    {
+      if (memcmp (current_p, import_name_p->u.char_p, current_length) == 0)
```
Merge these two `if` statements

LaszloLango commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+  while (stored_imports != NULL)
+  {
+    if (stored_imports->script_path.length == module_node_p->script_path.length
+       && memcmp (stored_imports->script_path.value_p,
```
wrong indentation

LaszloLango commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+                                bool is_import_item) /**< given item is import item */
+{
+  if (is_import_item &&
+      parser_module_check_for_duplicates (context_p, module_node_p, import_name_p))
```
Move `&&` to this line

LaszloLango commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+
+  parser_module_context_t *parent_context_p = JERRY_CONTEXT (module_top_context_p);
+  if ((parent_context_p == NULL || parent_context_p->exports_p == NULL || parent_context_p->imports_p == NULL)
+    && module_context_p->exports_p != NULL)
```
wrong indentation

LaszloLango commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,616 @@
+
+    if (context_p->token.type == LEXER_RIGHT_BRACE
+        || (context_p->token.type == LEXER_LITERAL
+        && lexer_compare_raw_identifier_to_current (context_p, "from", 4)))
```
wrong indentation

LaszloLango commented on jerry-port/default/default-io.c:
```diff
@@ -98,3 +99,79 @@ jerry_port_log (jerry_log_level_t level, /**< message log level */
+ * @return the source of the file
+ */
+uint8_t *
+jerry_port_module_read_source (const char *file_name_p, /**< file name */
```
The documentation update is missing for the new port API functions. BTW, why are these needed? Might be better to put the port API changes into a separate PR.

vincedani replied:
Thanks for the review, I’ve updated the port API documentation. The port functions are necessary for the modules, because without these Jerry could not open custom files (source files). For this reason, the changes can’t be separated into another PR.

rtakacs commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,605 @@
+
+    int request_count = exports_p->module_request_count + module_node_p->module_request_count;
+
+    if (request_count < MAX_IMPORT_COUNT)
```
```c
if ((uint_16_t) request_count < MAX_IMPORT_COUNT)  // error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
```

rtakacs commented on jerry-core/parser/js/js-parser-module.c:
```diff
@@ -0,0 +1,605 @@
+      stored_imports->module_names_p = module_node_p->module_names_p;
+
+      int request_count = stored_imports->module_request_count + module_node_p->module_request_count;
+      if (request_count < MAX_IMPORT_COUNT)
```
```c
if ((uint_16_t) request_count < MAX_IMPORT_COUNT)  // error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
```

rtakacs commented:
Other notes:
  * **jerry-core/module/module.c** should be renamed to **module_import.c** (or something like this). An other **module.c** already exist in the **jerry-ext** folder.
The Makefile of the NuttX target doesn't like if a filename is used multiple times (see [nuttx-stm32f4/Makefile#L47](https://github.com/pando-project/jerryscript/blob/master/targets/nuttx-stm32f4/Makefile#L47) - All the archive files are copied from JerryScript to NuttX. The copied archives are extracted to get the object files. If there are two **module.o** files (in different archive files) the **(jerry-core)module.o** will be overwritten by the **(jerry-ext)module.o** file). This causes linker errors.  

  * The **targets/nuttx-stm32f4/jerry-main.c** implements **jerry-port-log**. This functions is also implemented in **jerry-port/default/default-io.c**. Since both files are compiled, there is a linker error that Travis shows. (I've removed **jerry-port-log** from **jerry-main.c** to solve the linker error.)

 * Test files contain paths (**tests/jerry/...**) that should be opened. These paths can be different on boards - e.g. **js-remote-test** mounts **tests/jerry/** folder as **/test/**.

vincedani commented:
> Other notes:
> 
> * **jerry-core/module/module.c** should be renamed to **module_import.c** (or something like this). An other **module.c** already exist in the **jerry-ext** folder.
>   The Makefile of the NuttX target doesn't like if a filename is used multiple times (see [nuttx-stm32f4/Makefile#L47](https://github.com/pando-project/jerryscript/blob/master/targets/nuttx-stm32f4/Makefile#L47) - All the archive files are copied from JerryScript to NuttX. The copied archives are extracted to get the object files. If there are two **module.o** files (in different archive files) the **(jerry-core)module.o** will be overwritten by the **(jerry-ext)module.o** file). This causes linker errors.
> * The **targets/nuttx-stm32f4/jerry-main.c** implements **jerry-port-log**. This functions is also implemented in **jerry-port/default/default-io.c**. Since both files are compiled, there is a linker error that Travis shows. (I've removed **jerry-port-log** from **jerry-main.c** to solve the linker error.)
> * Test files contain paths (**tests/jerry/...**) that should be opened. These paths can be different on boards - e.g. **js-remote-test** mounts **tests/jerry/** folder as **/test/**.

Thank you @rtakacs for your review. I've moved and renamed the `module.{h, c}` as you suggested.

The paths in the test files are adjusted to the `./tools/runners/run-test-suite.sh`, which runs on Travis too. I think we should figure out how to adjust these paths when the the _js-remote-test_ environment is running.

zherczeg commented on jerry-core/ecma/base/ecma-module.c:
```diff
@@ -0,0 +1,258 @@
+                   size_t path_size, /**< length of the path */
+                   const char *source_p, /**< module source */
+                   size_t source_size, /**< length of the source */
+                   parser_module_node_t *module_node_p) /**< module node */
```
Misaligned arguments

zherczeg commented on jerry-core/ecma/base/ecma-module.c:
```diff
@@ -0,0 +1,258 @@
+                                              path_length,
+                                              (const char *) buffer_p,
+                                              size,
+                                              current_p);
```
Ditto. Please fix all of them.
